aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,369 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+  if (a.scalar_type() == at::ScalarType::Float) {
+    constexpr std::array<float, total_n_degs> thetas_float = {
+      1.192092800768788e-07, // deg 1
+      5.978858893805233e-04, // deg 2
+      5.116619363445086e-02, // deg 4
+      5.800524627688768e-01, // deg 8
+      1.461661507209034e+00, // deg 12
+      3.010066362817634e+00  // deg 18
+    };
+    return mexp_impl<float>(a, thetas_float);
+  }
+  else { // if Double
+    constexpr std::array<double, total_n_degs> thetas_double = {
+      2.220446049250313e-16, // deg 1
+      2.580956802971767e-08, // deg 2
+      3.397168839976962e-04, // deg 4
+      4.991228871115323e-02, // deg 8
+      2.996158913811580e-01, // deg 12
+      1.090863719290036e+00  // deg 18
+    };
+    return mexp_impl<double>(a, thetas_double);
+  }
+}
+
+// Based on:
+//
+// Mathias, Roy. 
+// “A Chain Rule for Matrix Functions and Applications.”
+// SIAM J. Matrix Anal. Appl. 17 (1996): 610-620.
+//
+template <typename func_t>
+Tensor backward_analytic_function_of_a_matrix(
+    const Tensor& self, const Tensor& grad,
+    const func_t& function_of_a_matrix
+  ) {
+  auto self_transposed = self.transpose(-2, -1);
+  auto self_transposed_sizes = self_transposed.sizes().vec();
+  self_transposed_sizes[self.dim() - 2] <<= 1;
+  self_transposed_sizes[self.dim() - 1] <<= 1;
+
+  auto n = self_transposed.size(-1);
+  auto meta_grad = at::zeros(self_transposed_sizes, grad.options());
+  meta_grad.narrow(-2, 0, n).narrow(-1, 0, n).copy_(self_transposed);
+  meta_grad.narrow(-2, n, n).narrow(-1, n, n).copy_(self_transposed);
+  meta_grad.narrow(-2, 0, n).narrow(-1, n, n).copy_(grad);
+
+  auto grad_input = function_of_a_matrix(meta_grad)
+    .narrow(-2, 0, n).narrow(-1, n, n);
+  return grad_input;
+}
+
+};
+
+// Computes the matrix exponential for a given batch of squared matrices.
+// The implementaion is based on:
+//
+// Bader, P.; Blanes, S.; Casas, F.
+// Computing the Matrix Exponential with an Optimized Taylor Polynomial Approximation.
+// Mathematics 2019, 7, 1174.
+//
+Tensor matrix_exp(const Tensor& a) {
+  TORCH_CHECK(a.dim() >= 2 
+          && (at::isFloatingType(a.scalar_type()) 
+           || at::isComplexType(a.scalar_type())),
+              "matrix_exp(", a.scalar_type(), "{", a.sizes(), "}): expected a tensor "
+              "of floating types with dim at least 2");
+  TORCH_CHECK(a.size(-1) == a.size(-2),
+              "matrix_exp(", a.scalar_type(), "{", a.sizes(), "}): expected a tensor "
+              "of squared matrices");
+
+  if (a.size(-1) == 1) {
+    return a.exp();
+  }
+
+  if (a.dim() <= 2) {
+    return mexp(a);
+  }
+  else {
+    auto res = at::empty(a.sizes(), a.options());
+    batch_apply(
+      {res, a},
+      a.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+
+        out.copy_(mexp(in));
+      }
+    );
+    return res;
+  }
+}
+
+Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
```

Comment:
In general, the backward functions are added in `tools/autograd/`. Entry goes in `derivatives.yaml` and function definition in `templates/Functions.cpp`.

Reply:
There is `batch_apply` method in here which I use in both forward and backward.

Reply:
When I try to move things into `Functions.cpp`, I get:
```
../torch/csrc/autograd/generated/Functions.cpp: In member function 'virtual torch::autograd::variable_list torch::autograd::generated::MatrixExpBackward::apply(torch::autograd::variable_list&&)':
../torch/csrc/autograd/generated/Functions.cpp:5022:74: error: call of overloaded 'matrix_exp_backward(at::Tensor&, at::Tensor&)' is ambiguous
     auto grad_result = any_grad_defined ? (matrix_exp_backward(self, grad)) : Tensor();
                                                                          ^
../torch/csrc/autograd/generated/Functions.cpp:630:8: note: candidate: at::Tensor torch::autograd::generated::{anonymous}::matrix_exp_backward(const at::Tensor&, const at::Tensor&)
 Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
        ^~~~~~~~~~~~~~~~~~~
In file included from ../aten/src/ATen/ATen.h:16:0,
                 from ../torch/csrc/autograd/generated/Functions.h:5,
                 from ../torch/csrc/autograd/generated/Functions.cpp:10:
aten/src/ATen/Functions.h:414:19: note: candidate: at::Tensor at::matrix_exp_backward(const at::Tensor&, const at::Tensor&)
 CAFFE2_API Tensor matrix_exp_backward(const Tensor & self, const Tensor & grad);

```
Do you know why?

Reply:
Ah, I see. Probably, I cannot pass `const Tensor&`...

Reply:
Could you try rebuilding from scratch again? It seems to me that there could be multiple definitions.

Reply:
I tried and it did not work. You can see that one of the candidates does not have constant references, which is strange, as my definition does have it, and so are the backward methods in `Functions.cpp`...

Reply:
@albanD , could you give a hand?

Reply:
Problem solved, needed to remove `matrix_exp_backward` from the `native_functions.yaml`.

Reply:
Turns out `Functions.cpp` does not like having template-based functions.

Reply:
`Functions.cpp` functions are basically in an anonymous namespace in the same cpp file where the Backward functions are defined.
So if you have an entry in native_functions with the same name it will be ambiguous yes.

I never tried to use templates there though, not sure why it wouldn't work.